### PR TITLE
KAFKA-16824: Utils.getHost and Utils.getPort do not catch a lot of invalid host and ports.

### PR DESCRIPTION
**Pull request by frankvicky**
Modify regex of HOST_PORT_PATTERN to prevent malformed hosts and ports.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ private Utils() {}
 
     // This matches URIs of formats: host:port and protocol:\\host:port
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: nit: Is the comment correct?
`protocol:\\host:port` -> ` protocol://host:port` ?

**Review by showuon — COMMENTED** [review]
LGTM! Just one minor comment.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 1 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Thank you, this is a typo, I will fix it. 😄

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Failed tests are unrelated.

**Comment by zhaochun-ma** [comment]
@frankvicky looks like with this change, SASL_PLAINTEXT://localhost:50132 becomes invalid url because of `_` in the protocol, 
and it's a breaking change, is this expected?

**Comment by showuon** [comment]
@zhaochun-ma , nice find! I've created [KAFKA-16946](https://issues.apache.org/jira/browse/KAFKA-16946) for this issue. Are you interested in opening a PR to fix it?

**Comment by zhaochun-ma** [comment]
> @zhaochun-ma , nice find! I've created [KAFKA-16946](https://issues.apache.org/jira/browse/KAFKA-16946) for this issue. Are you interested in opening a PR to fix it?

i haven't worked on kafka before, might be better for someone else to pick it up

**Comment by frankvicky** [comment]
Hi @zhaochun-ma 
Thanks for your remind 😄
If you have free cycle, it would be nice you could have a PR to fix it

**Comment by cmccabe** [comment]
> @frankvicky looks like with this change, SASL_PLAINTEXT://localhost:50132 becomes invalid url because of _ in the protocol, and it's a breaking change, is this expected?

definitely a bug. we should fix or revert

**Pull request merged**